### DISCUSSION TITLE
Photonic crystal effective index from phase delay

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#2829] [Q&A] [answered]
Hi - I'm looking to calculate the effective index of a uniform photonic crystal using meep. The idea is to extract the phase difference between two adjacent cells separated by lattice constant _a_ (to remove the contribution from the periodic function of the lattice from the Bloch wave) and calculate the phase velocity. 

The setup is pictured below. The fields are probed after 5 cells to allow for "proper" Bloch wave propagation. Periodic boundary conditions are used with PMLs along the _x_. This unit cell design comes from a "supercollimator" paper by @stevengj and others.

![Picture2](https://github.com/NanoComp/meep/assets/52722332/e8999830-a57f-45d5-b063-83c8a6ac87d3)

The process seems straightforward enough and indeed does work sometimes. But other times I get nonsensical results, like n_eff < 1.0. I'm thinking the issue is how I calculate phase velocity from these data (vp = dx/dt = 1 / (t2-t1)) and compatibility with meep units (n = 1/vp, where c = 1).

Any thoughts or suggestions are much welcome! I have not tried MPB because I assume PWEM will have the "band folding" issue and I'll be forced to validate the results through some other method anyway.

**Comment by stevengj** [comment, marked as answer]
Much easier to use MPB and compute the dispersion relation and modes for a single unit cell, they tell you everything about propagation in infinite periodic systems. 

Note that phase velocity is not well defined in a periodic structure because the modes are an Fourier series with infinitely many phase velocity components (equivalently, the propagation constant is only defined modulo reciprocal lattice vectors). This has nothing to do with the numerical method, it’s a basic consequence of the physics of periodic systems. 

On the other hand, phase delay, as defined from the frequency derivative of the phase, is essentially just distance divided by group velocity in a periodic system, and group velocity is perfectly well defined. MPB can compute group velocity for you. 

So you may need to rethink what question you want to ask.

**Reply by stevengj** [reply]
Also, PML doesn’t do what you think it does in this case. You are basically terminating your crystal with semi-infinite air, so you will have a reflection from the crystal/air boundary 

And it’s not clear how you are exciting your Bloch wave in Meep. You can’t just stick in an arbitrary source because you will generate evanescent fields too.

**Reply by stevengj** [reply]
Another basic question is what do you want to use the “effective index” for? 

Often textbook formulas for homogeneous media need to be re-thought for periodic media, you can’t necessarily just plug in the group index for refractive index.

**Reply by dalarev** [reply]
Good point on the PML! Absorber looks like a better boundary condition (never used it, but based on the docs).

A lens-embedded photonic crystal (ref [here](https://opg.optica.org/ol/fulltext.cfm?uri=ol-47-12-2943&id=476407), pic below) comprises two different cell designs, one for a photonic crystal operating under self-collimation, and the other serving as the "lens".

Calculating an effective index for both regions would allow treating these as a "conventional" lens with n_medium and n_lens such that we could use equations for focal length, etc., to compare the performance...at least that's the idea! It's sounding like even if you can extract an n_eff for one k & omega, it may not be generally useful.

![Screenshot 2024-05-09 165212](https://github.com/NanoComp/meep/assets/52722332/e006595a-0c15-4a8f-890e-ee61f678fc6a)

**Reply by stevengj** [reply]
I don’t think the conventional lens formulas are applicable to a PhC like this. Unless you’re operating at a very long wavelength, the propagation is quite different from a homogeneous medium. For example, the isofrequency surface in Fourier space is not a sphere. And at interfaces the transmission/reflection are much more complicated than the Fresnel equations (and can include multiple diffraction orders, or even a continuum of diffracted beams for interfaces at irrational slopes).

(This is the fundamental difficulty of trying to define a meaningful “effective index” for a PhC.)
